File: src/strategy/prosperitygrowthstrategy.cpp
#include "prosperitygrowthstrategy.h"

#include <algorithm>
#include <numeric>

namespace {
constexpr int MaxHistory = 180;
constexpr double PullbackTolerance = 0.015;
constexpr double ShrinkVolumeRatio = 0.90;
constexpr double BreakVolumeRatio = 1.15;
}

ProsperityGrowthStrategy::ProsperityGrowthStrategy(QObject *parent)
    : StrategyBase(parent)
    , m_partialTaken(false)
{
}

void ProsperityGrowthStrategy::reset()
{
    StrategyBase::reset();
    m_priceHistory.clear();
    m_closePrices.clear();
    m_volumes.clear();
    m_partialTaken = false;
}

void ProsperityGrowthStrategy::processMarketData(const MarketData& data)
{
    if (data.symbol != m_symbol || data.close <= 0.0) {
        return;
    }

    appendData(data);
    if (m_closePrices.size() < 60) {
        return;
    }

    const double ma20 = simpleMA(20);
    const double ma60 = simpleMA(60);
    const double prevMA60 = previousSimpleMA(60);
    const double avgVolume20 = averageVolume(20);

    if (m_hasPosition) {
        checkExitRules(data, ma60, avgVolume20);
        return;
    }

    checkEntryRules(data, ma20, ma60, prevMA60, avgVolume20);
}

void ProsperityGrowthStrategy::appendData(const MarketData& data)
{
    m_priceHistory.push_back(data);
    m_closePrices.push_back(data.close);
    m_volumes.push_back(data.volume);

    while (m_priceHistory.size() > MaxHistory) {
        m_priceHistory.pop_front();
        m_closePrices.erase(m_closePrices.begin());
        m_volumes.erase(m_volumes.begin());
    }
}

double ProsperityGrowthStrategy::simpleMA(int period) const
{
    if (period <= 0 || m_closePrices.size() < static_cast<size_t>(period)) {
        return 0.0;
    }

    return std::accumulate(m_closePrices.end() - period, m_closePrices.end(), 0.0) / period;
}

double ProsperityGrowthStrategy::previousSimpleMA(int period) const
{
    if (period <= 0 || m_closePrices.size() <= static_cast<size_t>(period)) {
        return 0.0;
    }

    return std::accumulate(m_closePrices.end() - period - 1, m_closePrices.end() - 1, 0.0) / period;
}

double ProsperityGrowthStrategy::averageVolume(int period) const
{
    if (period <= 0 || m_volumes.size() < static_cast<size_t>(period)) {
        return 0.0;
    }

    return std::accumulate(m_volumes.end() - period, m_volumes.end(), 0.0) / period;
}

int ProsperityGrowthStrategy::recentPullbackCount() const
{
    int count = 0;
    for (int i = static_cast<int>(m_closePrices.size()) - 1; i > 0; --i) {
        if (m_closePrices[static_cast<size_t>(i)] < m_closePrices[static_cast<size_t>(i - 1)]) {
            ++count;
            continue;
        }
        break;
    }
    return count;
}

bool ProsperityGrowthStrategy::externalConditionsConfirmed() const
{
    const GrowthBuyConfig& growth = m_config.growthBuyConfig;
    return growth.profitGrowthConfirmed
        && growth.orderLandingConfirmed
        && growth.noPureConceptConfirmed;
}

bool ProsperityGrowthStrategy::hasEnabledTrack() const
{
    return !m_config.growthBuyConfig.enabledTracks.isEmpty();
}

void ProsperityGrowthStrategy::checkExitRules(const MarketData& data, double ma60, double avgVolume20)
{
    const RiskConfig& risk = m_config.riskConfig;
    const double lotSize = m_config.positionConfig.lotSize;
    const double profitPercent = m_entryPrice > 0.0 ? (data.close / m_entryPrice - 1.0) * 100.0 : 0.0;
    const bool volumeBreak = avgVolume20 <= 0.0 || data.volume >= avgVolume20 * BreakVolumeRatio;

    if (risk.breakMA60VolumeStopEnabled && ma60 > 0.0 && data.close < ma60 && volumeBreak) {
        m_hasPosition = false;
        m_partialTaken = false;
        emitSignal(SignalType::STOP_LOSS, data.close, lotSize,
                   QStringLiteral("景气成长：放量跌破 60 均线，模拟离场"));
        return;
    }

    if (profitPercent <= -risk.stopLossPercent) {
        m_hasPosition = false;
        m_partialTaken = false;
        emitSignal(SignalType::STOP_LOSS, data.close, lotSize,
                   QStringLiteral("景气成长：达到通用止损线，模拟离场"));
        return;
    }

    if (profitPercent >= risk.surgeTakeProfitPercent && !m_partialTaken) {
        m_partialTaken = risk.partialTakeProfitEnabled;
        const double sellVolume = risk.partialTakeProfitEnabled ? lotSize / 2.0 : lotSize;
        if (!risk.partialTakeProfitEnabled) {
            m_hasPosition = false;
        }
        emitSignal(SignalType::TAKE_PROFIT, data.close, sellVolume,
                   QStringLiteral("景气成长：短期涨幅触发分批止盈"));
        return;
    }

    if (profitPercent >= risk.takeProfitPercent && !risk.partialTakeProfitEnabled) {
        m_hasPosition = false;
        emitSignal(SignalType::TAKE_PROFIT, data.close, lotSize,
                   QStringLiteral("景气成长：达到通用止盈线，模拟离场"));
    }
}

void ProsperityGrowthStrategy::checkEntryRules(const MarketData& data, double ma20, double ma60, double prevMA60, double avgVolume20)
{
    const GrowthBuyConfig& growth = m_config.growthBuyConfig;
    const bool ma60Up = !growth.requireMA60Up || (ma60 > 0.0 && prevMA60 > 0.0 && ma60 >= prevMA60);
    const bool nearMA = !growth.requirePullbackToMA20OrMA60
        || (ma20 > 0.0 && data.close <= ma20 * (1.0 + PullbackTolerance))
        || (ma60 > 0.0 && data.close <= ma60 * (1.0 + PullbackTolerance));
    const bool shrinkVolume = !growth.requireVolumePullback
        || avgVolume20 <= 0.0
        || data.volume <= avgVolume20 * ShrinkVolumeRatio;
    const int pullbackCount = recentPullbackCount();
    const bool pullbackWindow = pullbackCount >= growth.pullbackMinDays
        && pullbackCount <= growth.pullbackMaxDays;

    if (!hasEnabledTrack() || !externalConditionsConfirmed() || !ma60Up || !nearMA || !shrinkVolume || !pullbackWindow) {
        return;
    }

    m_hasPosition = true;
    m_partialTaken = false;
    m_entryPrice = data.close;
    m_entryTime = data.timestamp;
    m_stopLossPrice = ma60 > 0.0 ? ma60 : data.close * (1.0 - m_config.riskConfig.stopLossPercent / 100.0);
    m_takeProfitPrice = data.close * (1.0 + m_config.riskConfig.surgeTakeProfitPercent / 100.0);

    emitSignal(SignalType::BUY, data.close, m_config.positionConfig.lotSize,
               QStringLiteral("景气成长：人工确认项通过，技术回踩条件触发模拟分批低吸"));
}

File: quant/app/src/main/java/com/example/starquant/domain/strategy/ProsperityGrowthStrategy.kt
package com.example.starquant.domain.strategy

import com.example.starquant.data.model.MarketData
import com.example.starquant.data.model.SignalType
import java.util.ArrayDeque

class ProsperityGrowthStrategy : StrategyBase() {
    companion object {
        private const val MaxHistory = 180
        private const val PullbackTolerance = 0.015
        private const val ShrinkVolumeRatio = 0.90
        private const val BreakVolumeRatio = 1.15
    }

    private val priceHistory = ArrayDeque<MarketData>()
    private val closePrices = mutableListOf<Double>()
    private val volumes = mutableListOf<Double>()
    private var partialTaken: Boolean = false

    override fun reset() {
        super.reset()
        priceHistory.clear()
        closePrices.clear()
        volumes.clear()
        partialTaken = false
    }

    fun hasPositionState(): Boolean = hasPosition
    fun getEntryPrice(): Double = positionEntryPrice
    fun getStopLossPrice(): Double = positionStopLossPrice
    fun getTakeProfitPrice(): Double = positionTakeProfitPrice

    override fun processMarketData(data: MarketData) {
        if (data.symbol != symbol || data.close <= 0.0) return

        appendData(data)
        if (closePrices.size < 60) return

        val ma20 = simpleMA(20)
        val ma60 = simpleMA(60)
        val prevMA60 = previousSimpleMA(60)
        val avgVolume20 = averageVolume(20)

        if (hasPosition) {
            checkExitRules(data, ma60, avgVolume20)
            return
        }

        checkEntryRules(data, ma20, ma60, prevMA60, avgVolume20)
    }

    private fun appendData(data: MarketData) {
        priceHistory.addLast(data)
        closePrices.add(data.close)
        volumes.add(data.volume)

        while (priceHistory.size > MaxHistory) {
            priceHistory.removeFirst()
            closePrices.removeAt(0)
            volumes.removeAt(0)
        }
    }

    private fun simpleMA(period: Int): Double {
        if (period <= 0 || closePrices.size < period) return 0.0
        var sum = 0.0
        for (i in closePrices.size - period until closePrices.size) {
            sum += closePrices[i]
        }
        return sum / period
    }

    private fun previousSimpleMA(period: Int): Double {
        if (period <= 0 || closePrices.size <= period) return 0.0
        var sum = 0.0
        for (i in closePrices.size - period - 1 until closePrices.size - 1) {
            sum += closePrices[i]
        }
        return sum / period
    }

    private fun averageVolume(period: Int): Double {
        if (period <= 0 || volumes.size < period) return 0.0
        var sum = 0.0
        for (i in volumes.size - period until volumes.size) {
            sum += volumes[i]
        }
        return sum / period
    }

    private fun recentPullbackCount(): Int {
        var count = 0
        for (i in closePrices.size - 1 downTo 1) {
            if (closePrices[i] < closePrices[i - 1]) {
                count++
                continue
            }
            break
        }
        return count
    }

    private fun externalConditionsConfirmed(): Boolean {
        val growth = strategyConfig.growthBuyConfig
        return growth.profitGrowthConfirmed &&
                growth.orderLandingConfirmed &&
                growth.noPureConceptConfirmed
    }

    private fun hasEnabledTrack(): Boolean {
        return strategyConfig.growthBuyConfig.enabledTracks.isNotEmpty()
    }

    private fun checkExitRules(data: MarketData, ma60: Double, avgVolume20: Double) {
        val risk = strategyConfig.riskConfig
        val lotSize = strategyConfig.positionConfig.lotSize
        val profitPercent =
            if (positionEntryPrice > 0.0) (data.close / positionEntryPrice - 1.0) * 100.0 else 0.0
        val volumeBreak = avgVolume20 <= 0.0 || data.volume >= avgVolume20 * BreakVolumeRatio

        if (risk.breakMA60VolumeStopEnabled && ma60 > 0.0 && data.close < ma60 && volumeBreak) {
            hasPosition = false
            partialTaken = false
            emitSignal(
                SignalType.STOP_LOSS, data.close, lotSize,
                "景气成长：放量跌破 60 均线，模拟离场"
            )
            return
        }

        if (profitPercent <= -risk.stopLossPercent) {
            hasPosition = false
            partialTaken = false
            emitSignal(
                SignalType.STOP_LOSS, data.close, lotSize,
                "景气成长：达到通用止损线，模拟离场"
            )
            return
        }

        if (profitPercent >= risk.surgeTakeProfitPercent && !partialTaken) {
            partialTaken = risk.partialTakeProfitEnabled
            val sellVolume = if (risk.partialTakeProfitEnabled) lotSize / 2.0 else lotSize
            if (!risk.partialTakeProfitEnabled) {
                hasPosition = false
            }
            emitSignal(
                SignalType.TAKE_PROFIT, data.close, sellVolume,
                "景气成长：短期涨幅触发分批止盈"
            )
            return
        }

        if (profitPercent >= risk.takeProfitPercent && !risk.partialTakeProfitEnabled) {
            hasPosition = false
            emitSignal(
                SignalType.TAKE_PROFIT, data.close, lotSize,
                "景气成长：达到通用止盈线，模拟离场"
            )
        }
    }

    private fun checkEntryRules(
        data: MarketData,
        ma20: Double,
        ma60: Double,
        prevMA60: Double,
        avgVolume20: Double
    ) {
        val growth = strategyConfig.growthBuyConfig
        val ma60Up = !growth.requireMA60Up || (ma60 > 0.0 && prevMA60 > 0.0 && ma60 >= prevMA60)
        val nearMA = !growth.requirePullbackToMA20OrMA60 ||
                (ma20 > 0.0 && data.close <= ma20 * (1.0 + PullbackTolerance)) ||
                (ma60 > 0.0 && data.close <= ma60 * (1.0 + PullbackTolerance))
        val shrinkVolume = !growth.requireVolumePullback ||
                avgVolume20 <= 0.0 ||
                data.volume <= avgVolume20 * ShrinkVolumeRatio
        val pullbackCount = recentPullbackCount()
        val pullbackWindow = pullbackCount >= growth.pullbackMinDays &&
                pullbackCount <= growth.pullbackMaxDays

        if (!hasEnabledTrack() || !externalConditionsConfirmed() || !ma60Up ||
            !nearMA || !shrinkVolume || !pullbackWindow
        ) {
            return
        }

        hasPosition = true
        partialTaken = false
        positionEntryPrice = data.close
        entryTime = data.timestamp
        positionStopLossPrice =
            if (ma60 > 0.0) ma60 else data.close * (1.0 - strategyConfig.riskConfig.stopLossPercent / 100.0)
        positionTakeProfitPrice = data.close * (1.0 + strategyConfig.riskConfig.surgeTakeProfitPercent / 100.0)

        emitSignal(
            SignalType.BUY, data.close, strategyConfig.positionConfig.lotSize,
            "景气成长：人工确认项通过，技术回踩条件触发模拟分批低吸"
        )
    }
}
